### PR TITLE
Move service-url to config file

### PR DESCRIPTION
**Pull request by AnshulMalik**

### CODE REVIEW

**Comment by yurydelendik** [comment]
Fixes #23

It is still part of the bundle. Can we place it in the external file and read it using fetch()?

Sorry, I also did not mention is the issue, there is also service for the rust. Can you add this config to the PR?

**Comment by AnshulMalik** [comment]
Where should the external file exist? 

Don't know why this doesn't look right to me

**Comment by yurydelendik** [comment]
> Where should the external file exist?

Let's say we will name it e.g. "wasmservices.config" and place in the same folder as "index.html", e.g. in JSON format. The ts code will read this file asynchronously, probably cache, and give service address to the service.ts (via Promise/async).

File: config.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+  "serviceUrl": "http://wasmexplorer-service.herokuapp.com/service.php",
```
Comment (yurydelendik) [inline config.json]: start URL with '//' to not specify a schema

(you can modify the one below too)

File: src/service.ts [hunk 2]
```diff
@@ -93,12 +93,27 @@ export interface IServiceRequest {
...
+}
+
 export class Service {
-  static async sendRequest(command: string): Promise<IServiceRequest> {
-    const response = await fetch("//wasmexplorer-service.herokuapp.com/service.php", {
+  static config: IConfig;
+
+  static async sendRequest(body: string, to: string): Promise<IServiceRequest> {
+    if (!this.config) {
+      const configUrl = `${location.protocol}//${location.hostname}${(location.port ? ":" + location.port : "")}/config.json`;
```
Comment (yurydelendik) [inline src/service.ts]: define `configUrl` as './config.json'

File: src/service.ts [hunk 3]
```diff
@@ -93,12 +93,27 @@ export interface IServiceRequest {
...
+      this.config = await fetch(configUrl).then(resp => resp.json());
```
Comment (yurydelendik) [inline src/service.ts]: Extract `IConfig` and `fetch(configUrl).then(resp => resp.json())` into separate .ts module.

File: src/service.ts [hunk 4]
```diff
@@ -189,19 +204,13 @@ export class Service {
...
     } else if (from === Language.Rust && to === Language.Wasm) {
       // TODO: Temporary until we integrate rustc into the service.
-      return fetch("https://rust-heroku.herokuapp.com/rustc", {
-        method: "POST",
-        body: JSON.stringify({code: src}),
-        headers: new Headers({
-          "Content-Type": "application/json"
-        })
-      }).then(res => res.json());
+      return this.sendRequest(JSON.stringify({code: src}), "rustc");
```
Comment (yurydelendik) [inline src/service.ts]: It's hard to follow what `sendRequest` returns, in this case an object expected, however it was changed to return string.

Can you:
 - remove inaccurate refactoring and keep it as it was before
 - or, introduce sendRequestJSON to handle a new way of communicating with service?

File: src/service.ts [hunk 5]
```diff
@@ -189,19 +204,13 @@ export class Service {
         ]
       };
       const input = encodeURIComponent(JSON.stringify(project)).replace("%20", "+");
-      return this.sendRequest("input=" + input + "&action=build");
+      return this.sendRequest("input=" + input + "&action=build", "service");
```
Comment (yurydelendik) [inline src/service.ts]: Instead of passing "rustc" or "service" as a string, please introduce an enumeration.

File: src/service.ts (hunk 4 above)
Comment (AnshulMalik) [reply src/service.ts]: `sendRequest` return `return JSON.parse(await response.text());` which will be an object

File: src/config.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+const configUrl = "./config.json";
+
+export default async function getConfig() {
+  if (!this.config) {
+    this.config = await fetch(configUrl).then(resp => resp.json());
+  }
+
+  return this.config;
```
Comment (yurydelendik) [inline src/config.ts]: what is `this` in this context? for a module, you can declare the variable above

File: src/service.ts [hunk 7]
```diff
@@ -93,12 +94,32 @@ export interface IServiceRequest {
...
+
+    if (to === ServiceTypes.rustc) {
+      body = JSON.stringify(content);
+      url = config.rustc;
+      contentType = "application/json";
+    } else {
+      body = content.toString();
+      url = config.serviceUrl;
+      contentType = "application/x-www-form-urlencoded";
+    }
```
Comment (yurydelendik) [inline src/service.ts]: It's hard to the code logic. Could you create two methods `static async sendRequest(content: string | Object, to: ServiceTypes): Promise<IServiceRequest>` and `static async sendRequestJSON(content: string | Object, to: ServiceTypes): Promise<IServiceRequest>` to remove logic around assigning body/url/contentType variables and reduce complexity for simple method?

Comment (AnshulMalik) [reply src/service.ts]: will do

File: src/service.ts [hunk 8]
```diff
@@ -93,14 +94,34 @@ export interface IServiceRequest {
   output: string;
 }
 
+export enum ServiceTypes {
+  rustc,
```
Comment (yurydelendik) [inline src/service.ts]: nit: make enumeration values in Pascal case.

File: src/service.ts [hunk 9]
```diff
@@ -93,14 +94,34 @@ export interface IServiceRequest {
...
+      headers: new Headers({ "Content-Type": "application/json" })
     });
-    return JSON.parse(await response.text());
+
+    return response.json();
+  }
+
+  static async sendRequest(content: string, to: ServiceTypes): Promise<IServiceRequest> {
+    const config = await getConfig();
+    const url = (to === ServiceTypes.rustc) ? config.rustc : config.serviceUrl;
```
Comment (yurydelendik) [inline src/service.ts]: nit: no need for parens `()` in this expression

File: src/config.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
+export interface IConfig {
```
Comment (yurydelendik) [inline src/config.ts]: License header is missing

**Comment by yurydelendik** [comment]
Thank you for the patch
